### PR TITLE
Workflows to produce Python wheels

### PR DESCRIPTION
**Pull request by bettinaheim**
This PR
- Adds the workflows to create, cache and push devdeps images for the manylinux dev environment used to build the python wheels for linux. They are using the exact same workflow as the other devdeps images.
- Adds a matrix job to build and test python wheels during ci, using the wheel devdeps image(s).
- Adds a matrix job to do the same during deployment, only for a wider range of python versions, and potentially pushing to test-pypi in the future.
- Uploads the created wheels as artifacts both for ci and for deployments.
- Modifies the publishing workflow to grab the wheels produced during deployment, and upload them as assets if a release is created/preserved.
- Modifies the images and build scripts to build openblas and openssl from source. OpenBLAS is not yet built with OpenMP support; I haven't figured out how to build clang from source with OpenMP support (merely installing libomp doesn't suffice for cmake to find it)
- Configures the CPR build to build curl if no curl is found, and thus removes a system wide curl installation as a prerequisite for building the rest client. 
- Exludes the chemistry library from the build when the cpp frontend is disabled; We exclude the chemistry library from the python wheels because the chemistry library requires linking against the python version. See also [PEP513](https://peps.python.org/pep-0513/#libpythonx-y-so-1).
- Exludes tools and (unused) bin items from the python wheel to reduce size. 
- Reenables warnings as errors for the python builds.
- Builds ensmallen with OpenMP support if OpenMP is found.

### CODE REVIEW

File: CMakeLists.txt
Comment (amccaskey) [inline CMakeLists.txt:136]: Can we keep this openssl static variable as a CMake flag instead of hard-coded? I think this would make things harder on developers that just have apt-get install curl4-openssl-dev

File: CMakeLists.txt
Comment (amccaskey) [inline CMakeLists.txt:372]: This implies that we now want the RestClient to always be built. If we have a simulation-only user, they might want to be able to build the stack without needing remote qpu invocation.

Comment (bettinaheim) [reply CMakeLists.txt]: Alright, I'll remove it. I understand the convenience. At the same time, I think it is a bit too easy to accidentally remove part of the compilation without it being caught by CI. I think we need to think about a way to have the CI fail if parts that shouldn't be disabled are disabled.

File: CMakeLists.txt
Comment (bettinaheim) [reply CMakeLists.txt]: I changed to be set only if it is not defined already, which makes it static by default but allows one to override that setting. However, I have to say that you might run into issues if you build against one version but then have it pick up another version dynamically. I haven't had a chance to look very closely, but you can give it a try if you feel like it.

File: .github/workflows/publishing.yml
Comment (anthony-santana) [inline .github/workflows/publishing.yml:51]: For my understanding, this is going to return a wheel for each python version 3.9-3.11?

File: docker/build/cudaq.wheel.Dockerfile
Comment (anthony-santana) [inline docker/build/cudaq.wheel.Dockerfile:27]: Have the issues with python 3.10 that you mentioned today been resolved?

File: docker/build/devdeps.manylinux.Dockerfile
Comment (anthony-santana) [inline docker/build/devdeps.manylinux.Dockerfile:21]: Is this set to gcc because of what you mentioned in workflows/deployments.yml ?

> toolchain: gcc12 # set by the dockerfile and not configurable

File: python/CMakeLists.txt
Comment (anthony-santana) [inline python/CMakeLists.txt:74]: I see the issue you mentioned about the chemistry package linking against the python version. What does that mean long term for including them in the wheels? Will they have to become their own package?

**Review by anthony-santana — COMMENTED** [review]
Just left several questions, none of which should block the PR. Happy to approve once the threads are resolved and the CI passes

File: python/tests/backends/test_Quantinuum.py
Comment (anthony-santana) [inline python/tests/backends/test_Quantinuum.py:1]: You're probably going to have to add this same try loop in [test_Quantinuum_LocalEmulation.py](https://github.com/NVIDIA/cuda-quantum/blob/b387305db745ca923cd2d547d263c6021bdbd29f/python/tests/backends/test_Quantinuum_LocalEmulation.py) as well. I believe that's why the validate steps of the CI are failing

File: .github/workflows/publishing.yml
Comment (bettinaheim) [reply .github/workflows/publishing.yml:51]: 3.8-3.11, yes (i.e. for every version listed in the deployment workflow).

File: docker/build/cudaq.wheel.Dockerfile
Comment (bettinaheim) [reply docker/build/cudaq.wheel.Dockerfile:27]: I am not sure what you referring to. I mentioned running into numpy installation issues with 3.12. My be related https://github.com/numpy/numpy/issues/23808

File: docker/build/devdeps.manylinux.Dockerfile
Comment (bettinaheim) [reply docker/build/devdeps.manylinux.Dockerfile:21]: Oh, I forgot to update that - using x86_64 as the toolchain identifier (used in cache keys and image tags) seems appropriate instead for the image. Set to gcc because that was what manylinux comes with, and I didn't look further.

File: python/CMakeLists.txt
Comment (bettinaheim) [reply python/CMakeLists.txt:74]: Even if they are their own package, we should look into not linking against the python libraries. I would need to better understand where that need is coming from to define a plan.

**Comment by anthony-santana** [comment]
It seems like the version is getting dropped off of the wheel names?
```
cuda_quantum-0.0.0-cp38-cp38-manylinux_2_28_x86_64.whl
```
I'm also getting incompatible platform errors when installing both, so I'm guessing something in the build or auditwheel is breaking?
```
root@af63ce732121:/workspaces# pip install cuda_quantum-0.0.0-cp38-cp38-manylinux_2_28_x86_64.whl 
ERROR: cuda_quantum-0.0.0-cp38-cp38-manylinux_2_28_x86_64.whl is not a supported wheel on this platform.
```

File: python/tests/backends/test_Quantinuum.py
Comment (bettinaheim) [reply python/tests/backends/test_Quantinuum.py:1]: You are correct, and I updated it. It is, however, unclear to me why the local emulation test has a dependency on the mock server - it shouldn't. But that's a question for another PR.

File: docker/build/cudaq.wheel.Dockerfile
Comment (anthony-santana) [reply docker/build/cudaq.wheel.Dockerfile:27]: Ah bad memory on my part 🤦 disregard

File: .github/workflows/docker_images.yml
Comment (bettinaheim) [inline .github/workflows/docker_images.yml:564]:
```suggestion
```

**Review by anthony-santana — APPROVED** [review]

**Pull request merged**
